rpc/lib/client/ws_client.go

```diff
@@ -441,7 +446,12 @@ func (c *WSClient) readRoutine() {
 			continue
 		}
 		c.Logger.Info("got response", "resp", response.Result)
-		c.ResultsCh <- *response.Result
+		// Combine a non-blocking read on writeRoutineQuit with a non-blocking write on ResultsCh to avoid blocking
```

Comment:
what about the `c.ErrorsCh`?

Reply:
Actually this reminds me of an issue with `WSClient`, which might make sense to do something about now. It really would be better if it just exposed a `chan RPCResponse` rather than filtering them into `ResultsCh` and `ErrorsCh` because:

1. It forces callers to maintain and synchronise two goroutines to read a single logical stream of messages.
2. It would be easy to accomplish the same as a caller by launching a goroutine that reads from a single `chan RPCResponse`
3. (And by far the most important) currently callers have no access to the request/response correlation ID `RPCResponse.ID` that is set by the client and is meant to allow the client to map the `Responses`.

There's actually only 13 usages of `ResultsCh` in the Tendermint codebase. Maybe now would be a good time to fix this issue and just provide a `ResponseCh`. I think I could be persuaded to do the work to port it over in the Tendermint codebase, but are there other users we need to worry about a breaking change with?

Reply:
can't agree more! I would be in favor of doing that. @ebuchman any objections?